File: src/context/builder.ts
import { existsSync, readFileSync, readdirSync } from "node:fs";
import path from "node:path";

import { composeSystemPrompt, SYSTEM_PROMPT } from "./prompts.js";
import { ContextMessage, estimateLooseMessageTokens, estimateMessageTokens, estimateMessagesTokens, estimateTextTokens } from "./token-counter.js";
import type { WebSearchResult } from "../search.js";
import { buildContextSearchSection } from "../search.js";

const DEFAULT_CONTEXT_WINDOW_TOKENS = 1_000_000;
const DEFAULT_RESPONSE_RESERVE_TOKENS = 64_000;
const DEFAULT_MAX_FILES = 12;
export interface BuildContextOptions {
  rootDir: string;
  historySummary?: string;
  instruction: string;
  candidates: string[];
  explicitPaths?: string[];
  maxWindowTokens?: number;
  maxFiles?: number;
  projectPrompt?: string;
  reservedResponseTokens?: number;
  searchResults?: WebSearchResult[];
  systemPrompt?: string;
  topLevelEntryLimit?: number;
}

export interface ContextFile {
  path: string;
  content: string;
  mode: "full" | "compact" | "outline" | "truncated";
  tokenEstimate: number;
}

export interface BuildContextResult {
  messages: ContextMessage[];
  files: ContextFile[];
  projectMetadata: string;
  tokenEstimate: number;
  maxPromptTokens: number;
  truncated: boolean;
}

interface FileDraft {
  path: string;
  explicit: boolean;
  rawText: string;
}

export function buildContext(options: BuildContextOptions): BuildContextResult {
  const rootDir = path.resolve(options.rootDir);
  const explicitPaths = new Set((options.explicitPaths ?? []).map(normalizeRelativePath));
  const maxFiles = options.maxFiles ?? DEFAULT_MAX_FILES;
  const uniqueCandidates = dedupe(
    options.candidates.map(normalizeRelativePath).slice(0, maxFiles)
  );
  const fileDrafts = uniqueCandidates
    .map((filePath) => loadFileDraft(rootDir, filePath, explicitPaths.has(filePath)))
    .filter((draft): draft is FileDraft => draft !== null);
  let projectMetadata = buildProjectMetadata(rootDir, options.topLevelEntryLimit ?? 12);
  const maxPromptTokens =
    (options.maxWindowTokens ?? DEFAULT_CONTEXT_WINDOW_TOKENS) -
    (options.reservedResponseTokens ?? DEFAULT_RESPONSE_RESERVE_TOKENS);

  const systemMessage: ContextMessage = {
    role: "system",
    content: composeSystemPrompt(options.systemPrompt ?? SYSTEM_PROMPT, options.projectPrompt)
  };
  let metadataMessage: ContextMessage = {
    role: "user",
    content: projectMetadata
  };
  let baseTokens = estimateMessagesTokens([systemMessage, metadataMessage]);

  if (baseTokens > maxPromptTokens) {
    projectMetadata = buildMinimalProjectMetadata(rootDir);
    metadataMessage = {
      role: "user",
      content: projectMetadata
    };
    baseTokens = estimateMessagesTokens([systemMessage, metadataMessage]);
  }

  const emptyTaskTokens = estimateMessageTokens({
    role: "user",
    content: buildTaskMessage(options.instruction, [], options.searchResults ?? [], options.historySummary)
  });
  const availableFileTokens = Math.max(0, maxPromptTokens - baseTokens - emptyTaskTokens);
  let files = renderFiles(fileDrafts, "full");

  if (availableFileTokens === 0) {
    files = [];
  } else if (estimateRenderedFilesTokens(files, options.instruction, options.searchResults ?? []) > availableFileTokens) {
    files = renderFiles(fileDrafts, "compact");
  }

  if (estimateRenderedFilesTokens(files, options.instruction, options.searchResults ?? []) > availableFileTokens) {
    files = renderFiles(fileDrafts, "outline");
  }

  if (estimateRenderedFilesTokens(files, options.instruction, options.searchResults ?? []) > availableFileTokens) {
    files = dropNonExplicitFiles(files, explicitPaths);
  }

  if (estimateRenderedFilesTokens(files, options.instruction, options.searchResults ?? []) > availableFileTokens) {
    files = truncateFilesToBudget(files, availableFileTokens, options.instruction);
  }

  if (estimateRenderedFilesTokens(files, options.instruction, options.searchResults ?? []) > availableFileTokens) {
    files = forceFitFilesToBudget(
      files,
      explicitPaths,
      availableFileTokens,
      options.instruction,
      options.searchResults ?? []
    );
  }

  const taskMessage: ContextMessage = {
    role: "user",
    content: buildTaskMessage(options.instruction, files, options.searchResults ?? [], options.historySummary)
  };
  const messages = [systemMessage, metadataMessage, taskMessage];
  const tokenEstimate = estimateMessagesTokens(messages);

  return {
    messages,
    files,
    projectMetadata,
    tokenEstimate,
    maxPromptTokens,
    truncated: files.some((file) => file.mode !== "full")
  };
}

function loadFileDraft(rootDir: string, filePath: string, explicit: boolean): FileDraft | null {
  const absolutePath = path.join(rootDir, filePath);

  if (!existsSync(absolutePath)) {
    return null;
  }

  return {
    path: filePath,
    explicit,
    rawText: readFileSync(absolutePath, "utf8")
  };
}

function buildProjectMetadata(rootDir: string, topLevelEntryLimit: number): string {
  const entries = readdirSync(rootDir, { withFileTypes: true })
    .filter((entry) => ![".git", "node_modules", "dist", "build"].includes(entry.name))
    .sort((left, right) => left.name.localeCompare(right.name))
    .slice(0, topLevelEntryLimit)
    .map((entry) => `${entry.isDirectory() ? "dir" : "file"}: ${entry.name}`);

  return [
    "项目元信息：",
    `- 根目录: ${path.basename(rootDir)}`,
    "- 顶层结构：",
    ...entries.map((entry) => `  - ${entry}`)
  ].join("\n");
}

function buildMinimalProjectMetadata(rootDir: string): string {
  return ["项目元信息：", `- 根目录: ${path.basename(rootDir)}`].join("\n");
}

function renderFiles(fileDrafts: FileDraft[], mode: ContextFile["mode"]): ContextFile[] {
  return fileDrafts.map((draft) => {
    const content =
      mode === "full"
        ? draft.rawText
        : mode === "compact"
          ? compactText(draft.rawText)
          : mode === "outline"
            ? outlineText(draft.rawText)
            : draft.rawText;

    return {
      path: draft.path,
      content,
      mode,
      tokenEstimate: estimateTextTokens(content)
    };
  });
}

function compactText(text: string): string {
  return text
    .split(/\r?\n/u)
    .map((line) => line.trimEnd())
    .filter((line) => {
      const trimmed = line.trim();

      if (trimmed.length === 0) {
        return false;
      }

      return !(
        trimmed.startsWith("//") ||
        trimmed === "/*" ||
        trimmed === "*/" ||
        trimmed.startsWith("*")
      );
    })
    .join("\n");
}

function outlineText(text: string): string {
  const lines = compactText(text).split(/\r?\n/u);
  const structural = lines.filter((line) => {
    const trimmed = line.trim();

    return (
      trimmed.startsWith("import ") ||
      trimmed.startsWith("export ") ||
      trimmed.startsWith("class ") ||
      trimmed.startsWith("function ") ||
      trimmed.startsWith("interface ") ||
      trimmed.startsWith("type ") ||
      trimmed.startsWith("enum ") ||
      trimmed.startsWith("const ") ||
      trimmed.startsWith("let ") ||
      trimmed.startsWith("var ") ||
      trimmed.startsWith("async ") ||
      trimmed.includes("=>") ||
      trimmed.endsWith("{")
    );
  });

  const selectedLines = structural.length > 0 ? structural : lines.slice(0, 20);

  return selectedLines.join("\n");
}

function dropNonExplicitFiles(files: ContextFile[], explicitPaths: Set<string>): ContextFile[] {
  const kept: ContextFile[] = [];

  for (const file of files) {
    if (explicitPaths.has(file.path)) {
      kept.push(file);
    }
  }

  if (kept.length > 0) {
    return kept;
  }

  return files.slice(0, 1);
}

function truncateFilesToBudget(
  files: ContextFile[],
  availableFileTokens: number,
  instruction: string
): ContextFile[] {
  if (files.length === 0) {
    return files;
  }

  const instructionTokens = estimateTextTokens(instruction);
  const perFileBudget = Math.floor(Math.max(availableFileTokens - instructionTokens, 0) / files.length);

  return files.map((file) => {
    if (file.tokenEstimate <= perFileBudget) {
      return file;
    }

    return truncateFile(file, perFileBudget);
  });
}

function forceFitFilesToBudget(
  files: ContextFile[],
  explicitPaths: Set<string>,
  availableFileTokens: number,
  instruction: string,
  searchResults: WebSearchResult[]
): ContextFile[] {
  let fitted = [...files];
  let maxIterations = fitted.length * 2 + 5;

  while (fitted.length > 0 && estimateRenderedFilesTokens(fitted, instruction, searchResults) > availableFileTokens) {
    if (maxIterations-- <= 0) {
      break;
    }

    const overflow = estimateRenderedFilesTokens(fitted, instruction, searchResults) - availableFileTokens;
    const dropIndex = findLastNonExplicitIndex(fitted, explicitPaths);

    if (dropIndex >= 0 && fitted.length > 1) {
      fitted = fitted.filter((_, index) => index !== dropIndex);
      continue;
    }

    const largestIndex = findLargestFileIndex(fitted);
    const targetTokens = Math.max(8, fitted[largestIndex].tokenEstimate - overflow - 8);
    const nextFile = truncateFile(fitted[largestIndex], targetTokens);

    if (nextFile.tokenEstimate >= fitted[largestIndex].tokenEstimate) {
      break;
    }

    fitted = fitted.map((file, index) => (index === largestIndex ? nextFile : file));
  }

  return fitted;
}

function findLargestFileIndex(files: ContextFile[]): number {
  let largestIndex = 0;

  for (let index = 1; index < files.length; index += 1) {
    if (files[index].tokenEstimate > files[largestIndex].tokenEstimate) {
      largestIndex = index;
    }
  }

  return largestIndex;
}

function findLastNonExplicitIndex(files: ContextFile[], explicitPaths: Set<string>): number {
  for (let index = files.length - 1; index >= 0; index -= 1) {
    if (!explicitPaths.has(files[index].path)) {
      return index;
    }
  }

  return -1;
}

function truncateFile(file: ContextFile, targetTokens: number): ContextFile {
  if (targetTokens <= 16) {
    const content = "[omitted by context builder due to prompt budget]";

    return {
      path: file.path,
      content,
      mode: "truncated",
      tokenEstimate: estimateTextTokens(content)
    };
  }

  const targetChars = Math.max(32, targetTokens * 4);
  const head = file.content.slice(0, Math.floor(targetChars * 0.6));
  const tail = file.content.slice(-Math.floor(targetChars * 0.3));
  const content = `${head}\n... [truncated by context builder] ...\n${tail}`;

  return {
    path: file.path,
    content,
    mode: "truncated",
    tokenEstimate: estimateTextTokens(content)
  };
}

function estimateRenderedFilesTokens(
  files: ContextFile[],
  instruction: string,
  searchResults: WebSearchResult[]
): number {
  return estimateMessageTokens({
    role: "user",
    content: buildTaskMessage(instruction, files, searchResults, undefined)
  });
}

function buildTaskMessage(
  instruction: string,
  files: ContextFile[],
  searchResults: WebSearchResult[],
  historySummary: string | undefined
): string {
  const sections = [
    "当前任务：",
    instruction,
    "",
    "相关文件："
  ];

  if (historySummary) {
    sections.push(historySummary);
    sections.push("");
  }

  for (const file of files) {
    sections.push(`--- FILE: ${file.path} [mode=${file.mode}] ---`);
    sections.push(file.content);
    sections.push(`--- END FILE: ${file.path} ---`);
    sections.push("");
  }

  if (searchResults.length > 0) {
    sections.push(buildContextSearchSection(searchResults));
  }

  return sections.join("\n");
}

function dedupe(values: string[]): string[] {
  return [...new Set(values)];
}

function normalizeRelativePath(filePath: string): string {
  return filePath.replace(/\\/gu, "/").replace(/^\.\/+/u, "");
}

export function compressHistoryMessages<T extends { content?: string | null; role: string }>(
  messages: T[],
  maxPromptTokens: number,
  keepRecentRounds: number = 3
): T[] {
  const fixedCount = Math.min(2, messages.length);
  const fixed = messages.slice(0, fixedCount);
  const history = messages.slice(fixedCount);

  if (history.length === 0) {
    return messages;
  }

  const fixedTokens = fixed.reduce((sum, msg) => sum + estimateLooseMessageTokens(msg), 0);

  if (fixedTokens >= maxPromptTokens) {
    return fixed;
  }

  const rounds: T[][] = [];
  let currentRound: T[] = [];

  for (const msg of history) {
    if (msg.role === "assistant" && currentRound.length > 0) {
      rounds.push(currentRound);
      currentRound = [];
    }

    currentRound.push(msg);
  }

  if (currentRound.length > 0) {
    rounds.push(currentRound);
  }

  if (rounds.length <= keepRecentRounds) {
    return messages;
  }

  const droppedCount = rounds.length - keepRecentRounds;
  const keptRounds = rounds.slice(-keepRecentRounds);
  const summaryMessage = {
    role: "user" as T["role"],
    content: `（此前省略 ${droppedCount} 轮工具调用对话）`
  } as T;
  const compressed = [...fixed, summaryMessage, ...keptRounds.flat()];
  const compressedTokens = compressed.reduce(
    (sum, msg) => sum + estimateLooseMessageTokens(msg),
    0
  );

  if (compressedTokens <= maxPromptTokens) {
    return compressed;
  }

  const budget = maxPromptTokens - fixedTokens - estimateLooseMessageTokens(summaryMessage);
  const perRoundBudget = Math.floor(budget / keptRounds.length);

  const trimmedRounds = keptRounds.map((round) => {
    const roundTokens = round.reduce((sum, msg) => sum + estimateLooseMessageTokens(msg), 0);

    if (roundTokens <= perRoundBudget) {
      return round;
    }

    return round.map((msg) => {
      const content = msg.content ?? "";

      if (content.length <= 100) {
        return msg;
      }

      const targetChars = Math.floor(perRoundBudget * 4 * 0.6);
      const truncated = content.slice(0, targetChars) + "\n... [truncated]";

      return { ...msg, content: truncated } as T;
    });
  });

  return [...fixed, summaryMessage, ...trimmedRounds.flat()];
}

File: internal/context/builder.go
package context

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const defaultContextWindowTokens = 1000000
const defaultResponseReserveTokens = 64000
const defaultMaxFiles = 12

type BuildOptions struct {
	RootDir                string
	HistorySummary         string
	Instruction            string
	Candidates             []string
	ExplicitPaths          []string
	MaxWindowTokens        int
	MaxFiles               int
	ProjectPrompt          string
	ReservedResponseTokens int
	SearchResults          []SearchResult
	SystemPrompt           string
	TopLevelEntryLimit     int
}

type SearchResult struct {
	Title   string `json:"title"`
	URL     string `json:"url"`
	Snippet string `json:"snippet"`
}

type ContextFile struct {
	Path          string `json:"path"`
	Content       string `json:"content"`
	Mode          string `json:"mode"`
	TokenEstimate int    `json:"tokenEstimate"`
}

type BuildResult struct {
	Messages        []Message      `json:"messages"`
	Files           []ContextFile  `json:"files"`
	ProjectMetadata string         `json:"projectMetadata"`
	TokenEstimate   int            `json:"tokenEstimate"`
	MaxPromptTokens int            `json:"maxPromptTokens"`
	Truncated       bool           `json:"truncated"`
}

type fileDraft struct {
	Path     string
	Explicit bool
	RawText  string
}

func Build(options BuildOptions) (*BuildResult, error) {
	rootDir, err := filepath.Abs(options.RootDir)
	if err != nil {
		return nil, err
	}

	maxFiles := options.MaxFiles
	if maxFiles <= 0 {
		maxFiles = defaultMaxFiles
	}

	explicitPaths := map[string]bool{}
	for _, path := range options.ExplicitPaths {
		explicitPaths[normalizeRelativePath(path)] = true
	}

	candidates := dedupe(normalizeRelativePaths(options.Candidates))
	if len(candidates) > maxFiles {
		candidates = candidates[:maxFiles]
	}

	drafts := make([]fileDraft, 0, len(candidates))
	for _, candidate := range candidates {
		draft, err := loadFileDraft(rootDir, candidate, explicitPaths[candidate])
		if err != nil {
			continue
		}
		drafts = append(drafts, draft)
	}

	entryLimit := options.TopLevelEntryLimit
	if entryLimit <= 0 {
		entryLimit = 12
	}
	projectMetadata := buildProjectMetadata(rootDir, entryLimit)

	window := options.MaxWindowTokens
	if window <= 0 {
		window = defaultContextWindowTokens
	}
	reserve := options.ReservedResponseTokens
	if reserve <= 0 {
		reserve = defaultResponseReserveTokens
	}
	maxPromptTokens := window - reserve

	systemPrompt := options.SystemPrompt
	if systemPrompt == "" {
		systemPrompt = SystemPrompt
	}
	systemMessage := Message{
		Role:    "system",
		Content: ComposeSystemPrompt(systemPrompt, strings.TrimSpace(options.ProjectPrompt)),
	}
	metadataMessage := Message{
		Role:    "user",
		Content: projectMetadata,
	}
	baseTokens := EstimateMessagesTokens([]Message{systemMessage, metadataMessage})
	if baseTokens > maxPromptTokens {
		projectMetadata = buildMinimalProjectMetadata(rootDir)
		metadataMessage.Content = projectMetadata
		baseTokens = EstimateMessagesTokens([]Message{systemMessage, metadataMessage})
	}

	emptyTaskTokens := EstimateMessageTokens(Message{
		Role:    "user",
		Content: buildTaskMessage(options.Instruction, nil, options.SearchResults, options.HistorySummary),
	})
	availableFileTokens := maxPromptTokens - baseTokens - emptyTaskTokens
	if availableFileTokens < 0 {
		availableFileTokens = 0
	}

	files := renderFiles(drafts, "full")
	if availableFileTokens == 0 {
		files = nil
	} else if estimateRenderedFilesTokens(files, options) > availableFileTokens {
		files = renderFiles(drafts, "compact")
	}
	if estimateRenderedFilesTokens(files, options) > availableFileTokens {
		files = renderFiles(drafts, "outline")
	}
	if estimateRenderedFilesTokens(files, options) > availableFileTokens {
		files = dropNonExplicitFiles(files, explicitPaths)
	}
	if estimateRenderedFilesTokens(files, options) > availableFileTokens {
		files = truncateFilesToBudget(files, availableFileTokens, options.Instruction)
	}
	if estimateRenderedFilesTokens(files, options) > availableFileTokens {
		files = forceFitFilesToBudget(files, explicitPaths, availableFileTokens, options)
	}

	taskMessage := Message{
		Role:    "user",
		Content: buildTaskMessage(options.Instruction, files, options.SearchResults, options.HistorySummary),
	}
	messages := []Message{systemMessage, metadataMessage, taskMessage}
	truncated := false
	for _, file := range files {
		if file.Mode != "full" {
			truncated = true
			break
		}
	}

	return &BuildResult{
		Messages:        messages,
		Files:           files,
		ProjectMetadata: projectMetadata,
		TokenEstimate:   EstimateMessagesTokens(messages),
		MaxPromptTokens: maxPromptTokens,
		Truncated:       truncated,
	}, nil
}

func loadFileDraft(rootDir string, filePath string, explicit bool) (fileDraft, error) {
	absolutePath := filepath.Join(rootDir, filepath.FromSlash(filePath))
	data, err := os.ReadFile(absolutePath)
	if err != nil {
		return fileDraft{}, err
	}
	return fileDraft{
		Path:     filePath,
		Explicit: explicit,
		RawText:  string(data),
	}, nil
}

func buildProjectMetadata(rootDir string, topLevelEntryLimit int) string {
	entries, err := os.ReadDir(rootDir)
	if err != nil {
		return buildMinimalProjectMetadata(rootDir)
	}

	filtered := make([]os.DirEntry, 0, len(entries))
	for _, entry := range entries {
		if entry.Name() == ".git" || entry.Name() == "node_modules" || entry.Name() == "dist" || entry.Name() == "build" {
			continue
		}
		filtered = append(filtered, entry)
	}
	sort.Slice(filtered, func(i, j int) bool {
		return filtered[i].Name() < filtered[j].Name()
	})
	if len(filtered) > topLevelEntryLimit {
		filtered = filtered[:topLevelEntryLimit]
	}

	lines := []string{
		"Project metadata:",
		"- Root directory: " + filepath.Base(rootDir),
		"- Top-level structure:",
	}
	for _, entry := range filtered {
		kind := "file"
		if entry.IsDir() {
			kind = "dir"
		}
		lines = append(lines, "  - "+kind+": "+entry.Name())
	}
	return strings.Join(lines, "\n")
}

func buildMinimalProjectMetadata(rootDir string) string {
	return "Project metadata:\n- Root directory: " + filepath.Base(rootDir)
}

func renderFiles(drafts []fileDraft, mode string) []ContextFile {
	files := make([]ContextFile, 0, len(drafts))
	for _, draft := range drafts {
		content := draft.RawText
		switch mode {
		case "compact":
			content = compactText(draft.RawText)
		case "outline":
			content = outlineText(draft.RawText)
		}
		files = append(files, ContextFile{
			Path:          draft.Path,
			Content:       content,
			Mode:          mode,
			TokenEstimate: EstimateTextTokens(content),
		})
	}
	return files
}

func compactText(text string) string {
	var lines []string
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimRight(line, "\r\t ")
		trimmed := strings.TrimSpace(line)
		if trimmed == "" ||
			strings.HasPrefix(trimmed, "//") ||
			trimmed == "/*" ||
			trimmed == "*/" ||
			strings.HasPrefix(trimmed, "*") {
			continue
		}
		lines = append(lines, line)
	}
	return strings.Join(lines, "\n")
}

func outlineText(text string) string {
	lines := strings.Split(compactText(text), "\n")
	var structural []string
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "import ") ||
			strings.HasPrefix(trimmed, "export ") ||
			strings.HasPrefix(trimmed, "type ") ||
			strings.HasPrefix(trimmed, "interface ") ||
			strings.HasPrefix(trimmed, "class ") ||
			strings.HasPrefix(trimmed, "func ") ||
			strings.HasPrefix(trimmed, "function ") ||
			strings.HasPrefix(trimmed, "const ") ||
			strings.Contains(trimmed, "=>") ||
			strings.HasSuffix(trimmed, "{") {
			structural = append(structural, line)
		}
	}
	if len(structural) > 0 {
		return strings.Join(structural, "\n")
	}
	if len(lines) > 20 {
		lines = lines[:20]
	}
	return strings.Join(lines, "\n")
}

func dropNonExplicitFiles(files []ContextFile, explicitPaths map[string]bool) []ContextFile {
	var kept []ContextFile
	for _, file := range files {
		if explicitPaths[file.Path] {
			kept = append(kept, file)
		}
	}
	if len(kept) > 0 {
		return kept
	}
	if len(files) > 1 {
		return files[:1]
	}
	return files
}

func truncateFilesToBudget(files []ContextFile, availableFileTokens int, instruction string) []ContextFile {
	if len(files) == 0 {
		return files
	}

	instructionTokens := EstimateTextTokens(instruction)
	perFileBudget := (availableFileTokens - instructionTokens) / len(files)
	if perFileBudget < 0 {
		perFileBudget = 0
	}

	next := make([]ContextFile, 0, len(files))
	for _, file := range files {
		if file.TokenEstimate <= perFileBudget {
			next = append(next, file)
		} else {
			next = append(next, truncateFile(file, perFileBudget))
		}
	}
	return next
}

func forceFitFilesToBudget(files []ContextFile, explicitPaths map[string]bool, availableFileTokens int, options BuildOptions) []ContextFile {
	fitted := append([]ContextFile(nil), files...)
	maxIterations := len(fitted)*2 + 5
	for len(fitted) > 0 && estimateRenderedFilesTokens(fitted, options) > availableFileTokens {
		if maxIterations <= 0 {
			break
		}
		maxIterations--

		dropIndex := findLastNonExplicitIndex(fitted, explicitPaths)
		if dropIndex >= 0 && len(fitted) > 1 {
			fitted = append(fitted[:dropIndex], fitted[dropIndex+1:]...)
			continue
		}

		largest := findLargestFileIndex(fitted)
		nextFile := truncateFile(fitted[largest], fitted[largest].TokenEstimate/2)
		if nextFile.TokenEstimate >= fitted[largest].TokenEstimate {
			break
		}
		fitted[largest] = nextFile
	}
	return fitted
}

func truncateFile(file ContextFile, targetTokens int) ContextFile {
	if targetTokens <= 16 {
		content := "[omitted by context builder due to prompt budget]"
		file.Content = content
		file.Mode = "truncated"
		file.TokenEstimate = EstimateTextTokens(content)
		return file
	}

	targetChars := targetTokens * 4
	if targetChars < 32 {
		targetChars = 32
	}
	if len(file.Content) <= targetChars {
		return file
	}

	headLength := targetChars * 6 / 10
	tailLength := targetChars * 3 / 10
	content := file.Content[:headLength] + "\n... [truncated by context builder] ...\n" + file.Content[len(file.Content)-tailLength:]
	file.Content = content
	file.Mode = "truncated"
	file.TokenEstimate = EstimateTextTokens(content)
	return file
}

func findLastNonExplicitIndex(files []ContextFile, explicitPaths map[string]bool) int {
	for i := len(files) - 1; i >= 0; i-- {
		if !explicitPaths[files[i].Path] {
			return i
		}
	}
	return -1
}

func findLargestFileIndex(files []ContextFile) int {
	largest := 0
	for i := 1; i < len(files); i++ {
		if files[i].TokenEstimate > files[largest].TokenEstimate {
			largest = i
		}
	}
	return largest
}

func estimateRenderedFilesTokens(files []ContextFile, options BuildOptions) int {
	return EstimateMessageTokens(Message{
		Role:    "user",
		Content: buildTaskMessage(options.Instruction, files, options.SearchResults, options.HistorySummary),
	})
}

func buildTaskMessage(instruction string, files []ContextFile, searchResults []SearchResult, historySummary string) string {
	sections := []string{
		"Current task:",
		instruction,
		"",
		"Relevant files:",
	}

	if historySummary != "" {
		sections = append(sections, historySummary, "")
	}

	for _, file := range files {
		sections = append(sections, "--- FILE: "+file.Path+" [mode="+file.Mode+"] ---")
		sections = append(sections, file.Content)
		sections = append(sections, "--- END FILE: "+file.Path+" ---", "")
	}

	if len(searchResults) > 0 {
		sections = append(sections, "Web search results:")
		for _, result := range searchResults {
			sections = append(sections, "- "+result.Title+" "+result.URL+"\n  "+result.Snippet)
		}
	}

	return strings.Join(sections, "\n")
}

func dedupe(values []string) []string {
	seen := map[string]bool{}
	var result []string
	for _, value := range values {
		if value == "" || seen[value] {
			continue
		}
		seen[value] = true
		result = append(result, value)
	}
	return result
}

func normalizeRelativePaths(values []string) []string {
	result := make([]string, 0, len(values))
	for _, value := range values {
		result = append(result, normalizeRelativePath(value))
	}
	return result
}

func normalizeRelativePath(filePath string) string {
	path := filepath.ToSlash(filePath)
	return strings.TrimPrefix(path, "./")
}
